### PR TITLE
Prefix string literals with N in mssql

### LINKED ISSUE #3752: Need to add prefix 'N' before Unicode string when work with MSSQL
I am new to Sequelize and NodeJS, need to use it work with SQL Server 2014, and I also need to deal with Unicode string in my table columns.

In my model definition, I define a Name field with type DataTypes.STRING(255), the final real database column type is NVARCHAR(255). I can insert data with Unicode string as below sql statement:

    insert into [Test] values (N'我們')

Database can store the Chinese string correctly.

But when I use sequelize API, such as Model.create method, I will get question mark ( ???? ) in name column. After I check the NodeJS log, I find out the SQL statement generated by sequelize is as below:

    INSERT INTO [Test] ([Name]) OUTPUT INSERTED.* VALUES ('我們');

There is no prefix 'N' before the Chinese string, this is why it can't store correctly in Database.

I debug by myself but don't find a graceful way to change this behavior, is there any configuration not mentioned in the document?

I also submit a question in stackoverflow with more details (include my codes and my debug findings, also with an ugly temporary fix). Here is the link:
http://stackoverflow.com/questions/30303841/how-to-make-sequelize-add-n-before-unicode-string-in-its-sql-statement

### PR DESCRIPTION
**Pull request by zapatek**
Fixes #3752

### CODE REVIEW

**Comment by mickhansen** [comment]
This probably fits better than data types, @janmeier?

**Comment by zapatek** [comment]
@mickhansen, the enclosing quotes for string literals are put-in in the String.escape function. Since N'...' can be thought as a special variant of enclosing quotes,  I thought it would be the right function to modify.

On a separate note, this has been a showstopper for us. Not getting it fixed basically means sequelize is  not usable (with mssql) in majority of Asia. 

Given my unfamiliarity with internals of sequelize, If you could provide a more specific suggestion as to where to look at, I am willing to put in the effort and get this fixed at the soonest. On the other hand, if this is good enough then I can also fix those test cases which fail after this fix (they fail because they do not anticipate the N prefix in the built mssql queries at the moment).

**Comment by mickhansen** [comment]
@janmeier will chime have to chime in here, but since it's mssql specific i think it's better suited being implemented on mssql specific data types: https://github.com/sequelize/sequelize/blob/master/lib/dialects/mssql/data-types.js

**Comment by zapatek** [comment]
Thanks for the pointer, I'll look at it. Meanwhile, on the defence of the current approach, the modified `SqlString.escape` function has already some dialect specific code such as:

```javascript
  if (dialect === 'postgres' || dialect === 'sqlite' || dialect === 'mssql') {
    // http://www.postgresql.org/docs/8.2/static/sql-syntax-lexical.html#SQL-SYNTAX-STRINGS
    // http://stackoverflow.com/q/603572/130598
    val = val.replace(/'/g, "''");
  } else {
    val = val.replace(/[\0\n\r\b\t\\\'\"\x1a]/g, function(s) {
      switch (s) {
```

By the way, I have modified the test cases accordingly (anticipating the N prefix for mssql) and they all are passing now.

**Comment by mickhansen** [comment]
@zapatek It indeed does, and we'd like to fix that going forward :)

File: test/integration/model/find.test.js [hunk 1]
```diff
@@ -121,7 +121,7 @@ describe(Support.getTestDialectTeaser('Model'), function() {
           where: ['specialkey = ?', 'awesome'],
           logging: function(sql) {
             test = true;
-            expect(sql).to.contain("WHERE specialkey = 'awesome'");
+            expect(sql).to.match(/WHERE specialkey = N?'awesome'/);
```
Comment (mickhansen) [inline test/integration/model/find.test.js]: The `?` should not be present in the final result, as evident from the existing test.

**Comment by mickhansen** [comment]
Okay looking at the tests you changed i'm not sure that a change to data types will be sufficient actually, but we'll have to wait for @janmeier  to take a look

File: test/integration/model/find.test.js (hunk 1 above)
Comment (zapatek) [inline test/integration/model/find.test.js]: Ah, that question mark added in the fix is just a regular expression quantifier, meaning one or zero N characters are expected (to cover cases for mssql - one - and other dialects - zero - respectively)

Comment (mickhansen) [inline test/integration/model/find.test.js]: Oh right, my bad, for a second i missed that you chained it to a regex.

**Comment by janmeier** [comment]
Nah, I think this is the best way to fix it for now - either that or we have to move some more of the logic for proper escaping etc. into data types as well - So I think this is fine for now

**Comment by janmeier** [comment]
Could you add a changelog entry, then this should be good to go?

**Comment by zapatek** [comment]
Done.

**Comment by zapatek** [comment]
Seems like just missed the 3.15.0 release. Rebased the work branch for this fix to the tip of master to resolve the resultant conflict on changelog.md

**Comment by mickhansen** [comment]
I'll cut another one once i merge a few pending PRs :)

**Comment by mickhansen** [comment]
@zapatek Ah damn, looks like we need another rebase. Please do ping me when it's done and tests have run so i can get it merged in time.

**Comment by zapatek** [comment]
Amended the conflicting changelog.md but seems it wasn't in time for 3.15.1, so I put the entry under the 'Future' header. If you could merge now, it would surely make it into next release.. :)

**Comment by mickhansen** [comment]
And needs a new rebase, how the hell does this keep happening to this PR :p

**Comment by zapatek** [comment]
Haha, this time the culprit was `sql-string.js` changed by the recent commit 4f4abdc. I have just resolved the merge conflict. Could you try to merge the pr once again?

**Comment by brentjanderson** [comment]
This explains why my latin-encoded columns no longer work. I have a view that has the following structure:

Column_Name | Type | Length | Nullable | Collation
-------------------- | ------ | ---------| ------------- | ------------
Column_A	| char | 7 | no | SQL_Latin1_General_CP1_CI_AS
Column_B | char | 31 | no | SQL_Latin1_General_CP1_CI_AS

When I use a pretty standard Sequelize raw query, it generates the following with this new code:

```sql
SELECT [Column_A], [Column_B] 
FROM [xyzView] AS [xyzView] 
WHERE [xyzView].[Column_A] LIKE N'Value' OR [xyzView].[Column_B] LIKE N'Value'
```

Unfortunately, that code does not return any results. Dropping the Unicde prefix `N` from the strings in the `WHERE` clause does return results. I reviewed your changes, and it seems that all strings now are forced to unicode format via Sequelize. Did I miss an override? Should I get cracking on a pull request for such an option? Otherwise I may just pin my project at the older version.

**Comment by mickhansen** [comment]
@brentjanderson I don't believe any override is availab,e please do provide a PR

**Comment by Shaffuan** [comment]
Why don't update this changes in sequelize latest version???

**Pull request merged**